lms/views/lti/deep_linking.py

```diff
@@ -49,20 +62,96 @@
         },
     )
 
+    if application_instance.lti_version == "1.3.0":
+        context.js_config.add_deeplinking_api()
     return {}
+
+
+class DeepLinkingFieldsRequestSchema(JSONPyramidRequestSchema):
+    deep_linking_settings = fields.Dict(required=False, allow_none=True)
+    content_item_return_url = fields.Str(required=True)
+    content = fields.Dict(required=True)
+
+    extra_params = fields.Dict(required=False)
+
+
+@view_defaults(
+    request_method="POST", renderer="json", schema=DeepLinkingFieldsRequestSchema
+)
+class DeepLinkingFieldsViews:
+    """
+    Views that return the required form fields to complete a deeplinking request to the LMS.
+
+    After the user picks a document to be annotated the frontend will call these views
+    to get all the necessary fields to configure the assignment submitting them in a form to the LMS.
+    """
+
+    def __init__(self, request):
+        self.request = request
+
+        self.application_instance = self.request.find_service(
+            name="application_instance"
+        ).get_current()
+
+    @view_config(route_name="lti.deeplinking.formfields")
+    def file_picker_to_formfields_v13(self):
+        url = self._content_to_url(
+            self.request.route_url("lti_launches"),
+            self.request.parsed_params["content"],
+            self.request.parsed_params.get("extra_params"),
+        )
+
+        # In LTI1.3 there's just one `JWT` field which includes all the necessary information
+        return {
+            "JWT": self.request.find_service(LTIAHTTPService).sign(
+                {
+                    "https://purl.imsglobal.org/spec/lti/claim/deployment_id": self.application_instance.deployment_id,
+                    "https://purl.imsglobal.org/spec/lti/claim/message_type": "LtiDeepLinkingResponse",
+                    "https://purl.imsglobal.org/spec/lti/claim/version": "1.3.0",
+                    "https://purl.imsglobal.org/spec/lti-dl/claim/content_items": [
+                        {"type": "ltiResourceLink", "url": url}
+                    ],
+                    "https://purl.imsglobal.org/spec/lti-dl/claim/data": self.request.parsed_params[
+                        "deep_linking_settings"
+                    ],
+                }
+            )
+        }
+
+    @staticmethod
+    def _content_to_url(lti_launch_url, content, extra_params):
+        """
+        Translate content information from the frontend to a launch URL.
+
+        We submit the content information to the LMS as an URL pointing to our
+        `lti_launches` endpoint with any information required to identity
+        the content as query parameters.
+        """
+        params = dict(extra_params or {})
+
+        if content["type"] == "file":
+            params["canvas_file"] = "true"
+            params["file_id"] = content["file"]["id"]
+        else:
+            assert content["type"] == "url", "Unknown content type"
+            params["url"] = content["url"]
```

Comment:
I think the usage of assert here is hiding a coverage issue. We don't test the failure case here.

Reply:
Raising in that case now :+1: